fflinuxprint: init at 1.1.3-4

jaduff opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
New package - CUPS drivers for FujiFilm printers and multi-function printers.
###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jaduff committed:
maintainers: add jaduff

jaduff committed:
fflinuxprint: init at 1.1.3-4

drupol reviewed (changes requested):
Hi!

Thanks for your first contribution!

Can you please adjust your commit message an your PR title?

I invite you to have a look at [the contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) and reach out here if you need some assistance.

There should be 2 commits in this PR:

- 1 for adding you in the maintainer list: `maintainers: add jaduff`
- 1 for the rest: `fflinuxprint: init at 1.1.3-4`

Thanks!

jaduff commented:
Ok thanks for the feedback. I **think** I've fixed it. But I made some mistakes, and then went back and fixed them again. Please let me know if there's anything screwy!

drupol commented:
Don't forget to update the PR title. See others PRs to get your inspiration from !

drupol commented on pkgs/misc/cups/drivers/fflinuxprint/default.nix:
```diff
@@ -0,0 +1,68 @@
+    else if stdenv.hostPlatform.system == "i686-linux" then "i386"
+         else throw "Unsupported system: ${stdenv.hostPlatform.system}";
+in
+stdenv.mkDerivation rec {
```
I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

jaduff replied:
Ok I think I've done that.

drupol commented on pkgs/misc/cups/drivers/fflinuxprint/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+  src = fetchurl {
+    url = "https://support-fb.fujifilm.com/driver_downloads/fflinuxprint_1.1.3-4_amd64.deb";
+    sha256 = "434a81e20bc459ad74286b7a4a7815aab6b115e3f1210eb69d3ac5678e2fcab5";
```
This is optional since it could be automatized treewide at some point.

The idea is to avoid using `sha256` as attribute and use `hash` everywhere.

Since there could be different hashing algorithms in use, this make sense.

jaduff replied:
Done.

drupol commented on pkgs/misc/cups/drivers/fflinuxprint/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+  meta = with lib; {
+    description = "FujiFILM Linux Printer Driver";
+    longDescription = ''
```
I don't think we should get rid of all the supported devices or else we'll have to maintain it as well.

jaduff replied:
I've removed the details of the supported devices.

drupol commented on pkgs/misc/cups/drivers/fflinuxprint/default.nix:
```diff
@@ -0,0 +1,68 @@
+  dontConfigure = true;
+  dontBuild = true;
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out
+    mv etc $out
+    mv usr/lib $out
+    mkdir -p $out/share/cups/model
+    mv usr/share/ppd/fujifilm/* $out/share/cups/model
+    runHook postInstall
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall

    mkdir -p $out/share/cups/model
    mv {etc,usr/lib} $out
    mv usr/share/ppd/fujifilm/* $out/share/cups/model
    
    runHook postInstall
  '';
```

jaduff replied:
Changes made according to suggestion

jaduff commented:
I think I've made all the changes. Please let me know if I have done it right!

drupol commented:
@jaduff I force pushed a commit, it should fix all the remaining stuff.

drupol commented:
Thanks for your first contribution ! Keep PRs coming!
